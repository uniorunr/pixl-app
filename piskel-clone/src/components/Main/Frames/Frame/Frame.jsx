import React, { Component } from 'react';
import './Frame.scss';
import PropTypes from 'prop-types';
import { translateActiveFrame } from '../utils';

class Frame extends Component {
  componentDidMount = () => {
    const {
      duplicate, index, framesArray, resetDuplicate,
    } = this.props;
    if (duplicate) {
      const origin = framesArray[index - 1];
      const target = framesArray[index];
      const context = target.getContext('2d');
      context.imageSmoothingEnabled = false;
      context.drawImage(origin, 0, 0, target.width, target.height);
      translateActiveFrame(index);
      resetDuplicate();
    }
  };

  clickOnFrame = () => {
    const { index, makeActive } = this.props;
    makeActive(index);
    translateActiveFrame(index);
  };

  handleRemoveButton = () => {
    const { index, removeFrame } = this.props;
    removeFrame(index);
  };

  handleDuplicateButton = () => {
    const { index, duplicateFrame } = this.props;
    duplicateFrame(index);
  };

  setRef = (ref) => {
    const { innerRef } = this.props;
    this.ref = ref;
    innerRef(ref);
  };

  render() {
    const {
      active, index, framesArray, provided,
    } = this.props;

    return (
      <div
        role="presentation"
        data-id={index}
        data-active={active ? true : null}
        className={`frame frame${index}`}
        {...provided.draggableProps}
        {...provided.dragHandleProps}
        ref={this.setRef}
      >
        <canvas
          id={`frame${index}`}
          className={`frame__canvas ${active ? 'frame__canvas_active' : ''}`}
          width="128"
          height="128"
          onClick={this.clickOnFrame}
          ref={(ref) => {
            if (ref) {
              framesArray[index] = ref;
              framesArray.length = index + 1;
            }
          }}
        />
        <span className="frame__index">{index + 1}</span>
        <button
          className="frame__duplicate"
          type="button"
          onClick={this.handleDuplicateButton}
        >
          <i className="fas fa-copy" />
        </button>
        {index > 0 ? (
          <button
            className="frame__delete"
            type="button"
            onClick={this.handleRemoveButton}
          >
            <i className="fas fa-trash-alt" />
          </button>
        ) : null}
      </div>
    );
  }
}

Frame.propTypes = {
  index: PropTypes.number.isRequired,
  active: PropTypes.bool,
  removeFrame: PropTypes.func.isRequired,
  makeActive: PropTypes.func.isRequired,
  duplicateFrame: PropTypes.func.isRequired,
  resetDuplicate: PropTypes.func.isRequired,
  framesArray: PropTypes.instanceOf(Array),
  duplicate: PropTypes.bool,
  frameKeys: PropTypes.instanceOf(Array).isRequired,
  innerRef: PropTypes.func.isRequired,
  provided: PropTypes.instanceOf(Object).isRequired,
};

Frame.defaultProps = {
  active: false,
  framesArray: [],
  duplicate: false,
};

export default Frame;
