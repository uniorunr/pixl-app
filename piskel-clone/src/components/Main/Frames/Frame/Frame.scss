.frame {
  position: relative;

  .frame__canvas {
    width: 90px;
    height: 90px;
    border: 2px solid black;
    margin: 5px 0;
    box-sizing: border-box;
    cursor: pointer;
  }

  .frame__canvas_active {
    border: 3px dashed black;
  }

  .frame__index {
    position: absolute;
    top: 5px;
    left: 0;
    font-size: 1.2rem;
    background: white;
    padding: 4px 6px;
    border: 2px solid black;
  }

  .frame__duplicate {
    position: absolute;
    bottom: 8px;
    left: 0;
    border: 2px solid black;
    background: white;
    outline: none;
    cursor: pointer;
    font-size: 1.4rem;
    padding: 2px 5px;
  }

  .frame__delete {
    position: absolute;
    bottom: 8px;
    right: 0;
    border: 2px solid black;
    background: white;
    outline: none;
    cursor: pointer;
    font-size: 1.4rem;
    padding: 2px 5px;
  }
}
