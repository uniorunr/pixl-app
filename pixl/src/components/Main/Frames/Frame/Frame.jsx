import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import * as actions from '../../../../actions/actions';
import { translateActiveFrame, setActiveFrame } from '../utils';
import './Frame.scss';

class Frame extends Component {
  componentDidMount = () => {
    const {
      duplicate,
      index,
      framesArray,
      updateDuplicateFrameIndex,
    } = this.props;
    if (duplicate) {
      const origin = framesArray[index - 1];
      const target = framesArray[index];
      const context = target.getContext('2d');
      context.imageSmoothingEnabled = false;
      context.drawImage(origin, 0, 0, target.width, target.height);
      translateActiveFrame(index);
      updateDuplicateFrameIndex(null);
    }
  };

  clickOnFrame = () => {
    const { index, updateActiveFrameIndex } = this.props;
    updateActiveFrameIndex(index);
    translateActiveFrame(index);
  };

  handleRemoveButton = () => {
    const {
      index,
      frameKeys,
      activeFrame,
      updateActiveFrameIndex,
      updateFrameKeys,
    } = this.props;
    const indexToTranslate = setActiveFrame(activeFrame, frameKeys, index);
    frameKeys.splice(index, 1);
    translateActiveFrame(indexToTranslate);
    updateFrameKeys([...frameKeys]);
    const newActiveFrameIndex = indexToTranslate > frameKeys.length - 1
      ? frameKeys.length - 1
      : indexToTranslate;
    updateActiveFrameIndex(newActiveFrameIndex);
  };

  handleDuplicateButton = () => {
    const {
      index,
      frameKeys,
      updateDuplicateFrameIndex,
      updateActiveFrameIndex,
      updateFrameKeys,
    } = this.props;
    frameKeys.splice(index + 1, 0, Math.max(...frameKeys) + 1);
    updateFrameKeys([...frameKeys]);
    updateDuplicateFrameIndex(index + 1);
    updateActiveFrameIndex(index + 1);
  };

  setRef = (ref) => {
    const { innerRef } = this.props;
    this.ref = ref;
    innerRef(ref);
  };

  render() {
    const {
      active, index, framesArray, provided,
    } = this.props;

    return (
      <div
        role="presentation"
        data-id={index}
        data-active={active ? true : null}
        className={`frame frame${index}`}
        {...provided.draggableProps}
        {...provided.dragHandleProps}
        ref={this.setRef}
      >
        <canvas
          id={`frame${index}`}
          className={`frame__canvas ${active ? 'frame__canvas_active' : ''}`}
          width="128"
          height="128"
          onClick={this.clickOnFrame}
          ref={(ref) => {
            if (ref) {
              framesArray[index] = ref;
              framesArray.length = index + 1;
            }
          }}
        />
        <span className="frame__index">{index + 1}</span>
        <button
          className="frame__duplicate"
          type="button"
          onClick={this.handleDuplicateButton}
        >
          <i className="fas fa-copy" />
        </button>
        {index > 0 ? (
          <button
            className="frame__delete"
            type="button"
            onClick={this.handleRemoveButton}
          >
            <i className="fas fa-trash-alt" />
          </button>
        ) : null}
      </div>
    );
  }
}

Frame.propTypes = {
  index: PropTypes.number.isRequired,
  active: PropTypes.bool,
  activeFrame: PropTypes.number.isRequired,
  framesArray: PropTypes.instanceOf(Array).isRequired,
  duplicate: PropTypes.bool,
  frameKeys: PropTypes.instanceOf(Array).isRequired,
  innerRef: PropTypes.func.isRequired,
  provided: PropTypes.instanceOf(Object).isRequired,
  updateActiveFrameIndex: PropTypes.func.isRequired,
  updateDuplicateFrameIndex: PropTypes.func.isRequired,
  updateFrameKeys: PropTypes.func.isRequired,
};

Frame.defaultProps = {
  active: false,
  duplicate: false,
};

const mapStateToProps = state => ({
  framesArray: state.frames.framesArray,
  frameKeys: state.frames.frameKeys,
  activeFrame: state.frames.activeFrame,
});

const mapDispatchToProps = (dispatch) => {
  const {
    updateActiveFrameIndex,
    updateDuplicateFrameIndex,
    updateFrameKeys,
  } = bindActionCreators(actions, dispatch);
  return {
    updateActiveFrameIndex: (index) => {
      updateActiveFrameIndex(index);
    },
    updateDuplicateFrameIndex: (index) => {
      updateDuplicateFrameIndex(index);
    },
    updateFrameKeys: (keys) => {
      updateFrameKeys(keys);
    },
  };
};

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(Frame);
