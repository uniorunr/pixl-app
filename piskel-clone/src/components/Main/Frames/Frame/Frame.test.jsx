import React from 'react';
import renderer from 'react-test-renderer';
import Frame from './Frame';

const mockFunc = () => {};

it('renders correctly', () => {
  const tree = renderer
    .create(
      <Frame
        index={0}
        key={0}
        removeFrame={mockFunc}
        makeActive={mockFunc}
        active
        duplicateFrame={() => {}}
      />,
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
