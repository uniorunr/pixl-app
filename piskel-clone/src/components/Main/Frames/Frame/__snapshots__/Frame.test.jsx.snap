// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="frame frame0"
  data-active={true}
  data-id={0}
  role="presentation"
>
  <canvas
    className="frame__canvas frame__canvas_active"
    height="90"
    id="frame0"
    onClick={[Function]}
    width="90"
  />
  <span
    className="frame__index"
  >
    1
  </span>
  <button
    className="frame__duplicate"
    type="button"
  >
    <i
      className="fas fa-copy"
    />
  </button>
</div>
`;
